Added Cloning Content Guide

jyaistMap opened the pull request:
* draft to address cloning content guide [2230](https://github.com/ArcGIS/geosaurus/issues/2230)
 * please review to see if introduction makes sense - I attempted to put in a caveat about how cloning may not produce a one-to-one transfer (even though in the case of this notebook it did in my testing)
* I used the pythonapi playground and the geosaurus org. Both sets of items exist in the portals right now. If you want to run the notebook to test the results, perhaps you could create an admin account in geosaurus and run the notebook - to verify and test if the explanations are ok
* I did not include any real information about `related_items()`
-----

# Checklist

Please go through each entry in the below checklist and mark an 'X' if that condition has been met. Every entry should be marked with an 'X' to be get the Pull Request approved.


- [ ] All `import`s are in the first cell? First block of imports are standard libraries, second block are 3rd party libraries, third block are all `arcgis` imports? Note that in some cases, for samples, it is a good idea to keep the imports next to where they are used, particularly for uncommonly used features that we want to highlight.
- [ ] All `GIS` object instantiations are one of the following?
    - `gis = GIS()`
    - `gis = GIS('https://www.arcgis.com', 'arcgis_python', 'P@ssword123')`
    - `gis = GIS(profile="your_online_profile")`
    - `gis = GIS('https://pythonapi.playground.esri.com/portal', 'arcgis_python', 'amazing_arcgis_123')`
    - `gis = GIS(profile="your_enterprise_portal")`
- [ ] If this notebook requires setup or teardown, did you add the appropriate code to `./misc/setup.py` and/or `./misc/teardown.py`?
- [ ] If this notebook references any portal items that need to be staged on AGOL/Python API playground, did you coordinate with a Python API team member to stage the item the correct way with the api\_data\_owner user?
- [ ] Code refactored & split out across multiple cells, useful comments?
- [ ] Consistent voice/tense/narrative style? Thoroughly checked for typos?
- [ ] All images used like `<img src="base64str_here">` instead of `<img src="https://some.url">`? All map widgets contain a static image preview? (Call `mapview_inst.take_screenshot()` to do so)
- [ ] All file paths are constructed in an OS-agnostic fashion with `os.path.join()`? (Instead of `r"\foo\bar"`, `os.path.join(os.path.sep, "foo", "bar")`, etc.)
- [ ] **IF YOU WANT THIS SAMPLE TO BE DISPLAYED ON THE DEVELOPERS.ARCGIS.COM WEBSITE**, ping @ DavidJVitale so he can add it to the list for the next deploy

AtmaMani commented:
@CMPeng could you do the first round of reviews?

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>In paragraph 1, maybe to rewrite "The Python API team concluded we needed a guide for Cloning Content. We tackled the task to make the ultimate roadmap for how to get content from one Web GIS to another." from user's perspective?</p><p> </p><p>Then in paragraph 3, "Rather than tangle ourselves up when trying to distinguish between these portal types while writing about them, we'll refer to the portal or organization where the content originates as the <em>source</em> regardless of portal type. " - *tangling*, *ArcGIS organization*</p><p> </p><p>Also, in paragraph 4, "<code>Web Maps</code> consume these data service as. operational layer building blocks." - please remove the period.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

jyaistMap replied:
<ul><li>I edited the intro paragraph to read from the user perspective. How does this sound?:</li></ul><p> </p><p>				<em>One request the Python API team has heard repeatedly from Web GIS administrators: What's the best way </em></p><p>				<em>to move my content from a development Enterprise or organization to stagint to production?</em></p><p> </p><p>				<em>This Guide provides one possible roadmap for how to get content from one Web GIS to another - a template </em></p><p>				<em>for demonstrating basic software concepts when transferring content so the workflow can be modified for </em></p><p>				<em>other types of content and tailored for an administrator's particular need. So, first question to answer: What </em></p><p>				<em>does it mean to</em><code><em>clone</em></code><em>?.....</em></p><p> </p><ul><li>fixed the <code>tangling</code>  typo</li><li>remved extra period</li></ul>

chris-fox replied:
<p>Typo at "Cloning can refer to all kinds of transfer, from ArcGIS Enterprise to ArcGIS Online, Enterprise to Enterprise, or<strong> ArCGIS</strong> Online to Enterprise"</p>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>"Those items with a download, including file geodatabases, shapefiles, code samples, zip files, and packages amongst others" - Do you mean download-enabled items?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

jyaistMap replied:
<p>Changed it to read. I was under the impression that file types like this are automatically enabled for download. I clarified to refer to those types of items as separate for items that are enabled for export. Does this clarify?:</p><p> </p><p><em>Simple Types</em></p><ul><li><em>Those items with a download option (see&nbsp;</em><a href="https://developers.arcgis.com/rest/users-groups-and-items/items-and-item-types.htm" target="_blank"><em>Data files</em></a><em>&nbsp;for items that may be in a Web GIS and available for download), including zipped file geodatabases and shapefiles, code samples, zip files, and packages amongst others.</em></li></ul><p> </p>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>quotation marks missing on the right side -</p><p>should be <code>target = GIS(profile="your_online_admin_profile", verify_cert=False)</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

jyaistMap replied:
<p>Fixed it!</p>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>Just curious, can an administrator clone items belonging to another administrator? Also, is there any requirement for source administrator and target administrator to be having the same user name, and other license types/privileges/roles?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

jyaistMap replied:
<ul><li>I tested - logged in as one administrator, I could clone the items of another administrator.</li><li>There is no requirement for the usernames of the administrators to match. The source code gets username values form the gis object.</li><li>The source code checks to ensure the target user has these privileges depending upon the item type: <em>portal:user:createItem, portal:user:createGroup, or portal:publisher:publishFeatures.  </em>The clone_items() function will fail if the necessary privilege is not present when cloning the respective item type.</li></ul>

chris-fox replied:
<p>A user can clone any items they have access to. So if items are shared publicly i can clone them from any org to my org. We are just using the sharing model, so if it is shared with me I can clone it, but as John described I can't create anything that I am don't have the privileges to do. So if I can create hosted feature layers through the UI, I can't clone them.</p>

jyaistMap replied:
<p>Added a note in the guide to explain this:</p><p> </p><p><strong><em>NOTE:</em></strong><em>&nbsp;Cloning relies on the sharing model to determine the items any particular&nbsp;</em><code><a href="https://develoipers.arcgis.com/python/api-reference/arcgis.gis.toc.html#user" target="_blank"><em>user</em></a></code><em>&nbsp;can clone. If a user can access an item, that user can clone it. However, that user can't create anything in the target portal they don't have the appropriate&nbsp;</em><code><a href="https://enterprise.arcgis.com/en/portal/latest/administer/windows/roles.htm#ESRI_SECTION1_4FF9051EFB814C249AB26B0ACFF7C79F" target="_blank"><em>privileges</em></a></code><em>&nbsp;to create.</em></p>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>Can we have a newline before each line of the item type, to distinguish the item types (of the next category) and item names (of the current category)?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

jyaistMap replied:
<p>The display in ReviewNB does not preserve the formatting from the Notebook. I have the ouput grouped so there is a new line separating each Item Type group...and the individual items are listed with an indent. It is more legible in the notebook itself. I'll add a screenshot in GitHub. Let me know if that makes it clearer:</p><p> </p><p>Item Type</p><p>------------------</p><p>    Item1</p><p>    item2</p><p> </p><p>Item Type2</p><p>------------------ </p><p>    ItemType2_1</p><p>    ItemType2_2</p><p>    ItemType2_3</p><p> </p><p>user_inv</p>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>Can we confirm that <code>Hosted Service</code>  shows up in typeKeywords of feature services, regardless of versions of portal/arcgis online? If so, is it better to put down some descriptions of how "hosted service" in the keywords distinguish hosted FS from feature server extensions?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

jyaistMap replied:
<ul><li>I talked to the author of the clone_items function and the PM for ArcGIS Online, who both believe the <code>typeKeywords</code>  property is a reliable method for filtering our hosted features services. The other way would be to query the <code>serviceDefinition</code>   for the word 'Hosted'. Using the <code>item</code> is much simpler and straightforward. There may be edge cases if you add a hosted feature layer as an item or <em>Save as a Copy</em>. </li></ul>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>Can we confirm that <code>nested</code> folder structures are also created in this case?  e.g. a FS exists under folder_level_a\folder_level_b\folder_level_c?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

jyaistMap replied:
<p>The folder structure I'm duplicating is the folder structure in the portal app. I don't see how to create nested folders in the app. Do you do that through the REST API? I've not seen that. I can make a note that the code is only one level deep on folders. Let me know what you mean about the nested folders.</p>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>The step to validate if a web map is successfully cloned from source to target is great here. Is there a way to also find out that the web map can be opened correctly, and the links of the containing FS are not broken?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='64'/>

jyaistMap replied:
<p>I added code cells visualizing one of the source webmaps and the corresponding target webmaps. Does that verification seem solid?</p>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>MapView screenshot missing.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='71'/>

jyaistMap replied:
<p>Fixed it!</p>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>screenshot missing</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='82'/>

jyaistMap replied:
<p>Fixed it!</p>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>screenshot missing</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='92'/>

jyaistMap replied:
<p>Fixed it!</p>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>screenshot missing</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='105'/>

jyaistMap replied:
<p>Fixed it!</p>

PGCHM commented on guide/03-the-gis/cloning-content.ipynb:
```diff
@@ -0,0 +1,2051 @@
+{
```
<p>Thanks John for authoring this wonderful guide, it has been nicely structured and comprehensive, in presenting a universally workable approach for cloning items from source to target portals. Very neat, thanks!! </p><p>I have raised some questions in the previous comments, please check!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='108'/>

CMPeng commented:
> @CMPeng could you do the first round of reviews?

Sure @AtmaMani  Please see comments made above. Sorry - I think I forgot to switch account, and made comments with my personal GitHub account. Hope you dont mind @jyaistMap

jyaistMap commented:
<p>Thank you for the review <a href="https://github.com/CMPeng" target="_blank">@CMPeng</a> and the great suggestions and compliment. I just pushed some changes to the branch so take a look when you can.</p>

CMPeng commented:
Thanks @jyaistMap for the quick fixes. All looks good to me, except for one typo in the same sentence -
```
One request the Python API team has heard repeatedly from Web GIS administrators: What's the best way to move my content from a development Enterprise or organization to **_stagint_** to production?
```
Otherwise, all thumbs up. It is back to you, @AtmaMani

jyaistMap commented:
Thanks @CMPeng - fixed that typo!

AtmaMani commented:
@BP-Ent  could you review this PR? You need to use the "ReviewNB" link (can be found in the 1st comment by the bot) to see the notebook and post your comments.

BP-Ent commented:
@AtmaMani Hi Atma,
It looks like I can access the PR in ReviewNB, as well as comment. I'm not familiar with ReviewNB, so I'll need to get acquainted with it. After that, I can review the notebook and comment any suggested edits I find.

jyaistMap commented:
<p>Made the suggested changes,</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Done!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Done!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Done!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Done!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Done!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Thanks for the suggestions, <a href="https://github.com/BP-Ent" target="_blank">BP-Ent</a>! Made all the changes.</p>

jyaistMap commented:
Hey @BP-Ent Ignore any message you get from me about re-requesting a review on this. I pressed  an icon inadvertently.

jyaistMap commented:
@AtmaMani In talking with @chris-fox, I'm going to adjust some of the narrative to talk more about a couple parameters and adjust how I used the `item_mapping` in a couple places.

jyaistMap commented:
<p>Fixed this!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Added those item types to the list in the introduction.</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Changed the function to this:</p><p> </p><p><code></code><code>python</p><p>def get_layer_item_ids(wm):</p><p>&nbsp;&nbsp;wmo = WebMap(wm)</p><p>&nbsp;&nbsp;wm_id_list = []</p><p>&nbsp;&nbsp;for layer in wmo.layers:</p><p>&nbsp;&nbsp;&nbsp;&nbsp;try:</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;fsvc = FeatureLayerCollection(layer['url'][:-1], active_gis)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if not fsvc.properties['serviceItemId'] in wm_id_list:</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;wm_id_list.append(fsvc.properties['serviceItemId'])</p><p>&nbsp;&nbsp;&nbsp;&nbsp;except Exception as e:</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;pass</p><p>&nbsp;&nbsp;return wm_id_list</p><p></code><code></code><span class="ql-cursor">﻿</span></p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p><code>`</code>python</p><p> print("test")</p><pre class="ql-syntax" spellcheck="false">
</pre>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Thanks for the explanation <a href="https://github.com/chris-fox" target="_blank">chris-fox</a>. I've significantly edited that section to remove the cloning order and replaced it with explanations of the <em>search_existing_items</em> and <em>item_mapping</em> parameters.</p><p> </p><p> </p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAFEAAABRCAYAAACqj0o2AAAACXBIWXMAAAsTAAALEwEAmpwYAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAARuSURBVHgB7Z1bVtNQFIb/E7y8dgZ2BuIIrCMQRyCOQHhEdAELAd/AEYAjAEdAZ2CdQZxBfRXa497nJJCmSZOcXNom+1srbRra0vxrX07OZUdhmRzqHjYwgEKftpfmWdNm6cfePabNp/eM6T0+pvhDr4eYYIRDNcYSUWiSR9Fe07aFeaFc8cGCTvGTRB02LWozIh7rAYn3kSxoQK96qJ8rEvQHvqghGqA+EdnqnpNwU+ygGeGS8Gk7wh1u6rTO6kVcDfHi+PDIOvfUEWqgWhFP9DY9HqC6WFc1Ptgy99UVKqQaEc80Z9XLIOatPorc+x92ycV9VICHsnzT7Lq/1kZARlPL4Cn9Zus5pXG3RI59z8h1tYl964vCBT6pXZTATUR23ymuaW8T7WBEGfydq3sXF9EKeIvVTR6u+CTkGxchi4nYXgFDnITML2LTAu6r2d92ojWaobCQ+bJz+y0wSp8y9y0lzn7eD+QT0SaRProDC3ltWiA5yBbxVJ+jPVm4CJumCZeDxSJyY3Td24Fl4HM/1Znnny7imYkJ5+g6mqwxIz56Cz7MAq5KL8wy4Suzy0VvSBbRuvEWBAv3Cyy4zk6zxFwBtWMcpGXreRHP9Cr3By6TPg1xJCaZWRFZ6Sm2ISTjUY99gjXOivi00hG4NtJLssa4O0sszIKtce5QyLEWK8xHzwwBR/Aie28h5GNj1mOtiDZYbkPIh6br6kiCsSJurNEg02rQi2rmBY/iysUZhDte/ICQk4jheYFv9yEUpR/GRY98u4sdrtUQxEV25wEEN5T1YC/cERzg2b1gET28gOCGDi1RS++1Mw/uLJm5DMYAn6DoOEp8ZkJTNPV/i820CJo4QmlExAoQESuAY6KPIsmlqdlZy5sVVgSzrEMssRyBiMpYouCCttp5wUJDwQWFv/zE7jyC4MbUaueFJik4EYjIS1sFNyahiHb1pQ+hKH44Od42cXixtVCUYbjjxQ8IOYkYnhXRxsWl1lFYOyZxS+S4qKSpU4Cr6Er+x8u+CWpZld5KYjnkUURbdEJcOhuftLqJHpjtgJjiO4Qs5jx2VsQJLiDWuAhePDmMH4zN2aZgKdaYDtfaSVh9Ot+fKNaYBsfCw6Q/zItoU3epmggtJbX1ktyzzXVjlFzFPKBJiwW1dNKHBxQ+QNyaGePeaJFKuoh7FEC1NMCNBhmlDBYPVH1WF53O1nzurEEG2aN9Exyim0MIfnDumWSLyNnawzt0q+M2rEiSKyfkG3fm+OjRl3ZDyMIlXaS40Cw1FxcKkTJXcxSfRtJO1x65Csi4zcVhIe/wqhXNHz6HEgIy5WefftU79C286nLd5n6PTUM6Rzswi26WQ+VrYb6Uq6gcatcK89oeqpUszBvFVnXfoWj7Hqsjpu1s5r7SGupt1zcj3zaFBliuZdYqXkhTZfO3zL0GlLHOuuGEMTJJY+3L5ifxeAOHLXMTB7mBQwWwy9+bpcKbFENfmiVydqlXD/NNpjEQ3EaEp0hr/Db7SxBNqIH/xlx08qKeJuIAAAAASUVORK5CYII=</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Removed that part of the code and the explanation above. Replaced it with some code that displays the target item and it's source-typeKeyword and displays it with the source item and its idea.</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>I took out the <em>item_mapping </em>and replaced it with an explanation - and then compare the resulting ids of the items in the webmap.</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Thanks for this explanation. I saw in the <code>_clone</code>  code that <code>web maps</code>  are not automatically brought over in the classic story map.  <a href="https://github.com/chris-fox" target="_blank">chris-fox</a>, thanks again for looking through this and clarifying <code>item_mapping</code>.</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
@AtmaMani I talked to @chris-fox and made the suggested edits. So this guide is ready for further review.

jyaistMap commented:
<p>Made the edits!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Removed it!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Great!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Thanks for the tip! Made the change.</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Good catches! Corrected the spelling and changed the sentence to reference <code>wm_map</code>.</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Thanks. Removed it!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Good - hopefully it emphasizes the utility of <code>clone_items()</code>!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>Agreed!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
<p>That is truly a confusing sentence. Changed it to:</p><p> </p><p><em>The </em><code><em>clone_items()</em></code><em> function provides the means to migrate data service items and more complex items that consume them between Web GIS systems. For items that consume other items, the function provides parameters to control the items they consume.</em></p><p><em> </em></p><p>Does that read better and summarize the function?</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jyaistMap commented:
@AtmaMani I updated the guide with demonstrations immediately after the intro and used the `owner` parameter.

AtmaMani commented:
<p>How about:</p><p><em>The </em><code><em>clone_items()</em></code><em> function provides the means to migrate maps, apps, web layers and more complex items between Web GIS systems. For items that consume other items (such as a web map item consuming a set of web layer items), the function provides parameters to control how those get migrated.</em></p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/677/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AtmaMani commented:
@jyaistMap I reviewed the new addition (quick start example) and it looks good. I made a suggestion for the conclusion, once that's done, we can merge this PR.

jyaistMap commented:
Made that update, @AtmaMani. Thanks for the feedback.
